src/toil/bus.py

```diff
@@ -49,39 +204,198 @@ class MessageBus:
+
+    # This next function takes callables that take things of the type that was passed in as a
+    # runtime argument, which we can explain to MyPy using a TypeVar and Type[]
+    MessageType = TypeVar('MessageType', bound='NamedTuple')
+    def subscribe(self, message_type: Type[MessageType], handler: Callable[[MessageType], Any]) -> Listener:
+        """
+        Register the given callable to be called when messages of the given type are sent.
+        It will be called with messages sent after the subscription is created.
+        Returns a subscription object; when the subscription object is GC'd the subscription will end.
+        """
+
+        topic = self._type_to_name(message_type)
+        logger.debug('Listening for message topic: %s', topic)
 
-        # Log who sent the message, and what it is
-        # See: <https://stackoverflow.com/a/57712700>
-        our_frame = inspect.currentframe()
-        assert our_frame is not None, "Interpreter for Toil must have Python stack frame support"
-        caller_frame = our_frame.f_back
-        assert caller_frame is not None, "MessageBus.put() cannot determine its caller"
-        logger.debug('%s sent: %s', caller_frame.f_code.co_name, message)
+        # Make sure to wrap the handler so we get the right argument name and
+        # we can control lifetime.
+        def handler_wraper(message: MessageBus.MessageType) -> None:
+            handler(message)
 
-        self.__messages_by_type[type(message)].append(message)
+        # The docs says this returns the listener but really it seems to return
+        # a listener and something else.
+        listener, _ = self._pubsub.subscribe(handler_wraper, topic)
+        # Hide the handler function in the pubsub listener to keep it alive.
+        # If it goes out of scope the subscription expires, and the pubsub
+        # system only uses weak references.
+        setattr(listener, 'handler_wrapper', handler_wraper)
+        return listener
+
+    def connect(self, wanted_types: List[type]) -> 'MessageBusConnection':
+        """
+        Get a connection object that serves as an inbox for messages of the
+        given types.
+        Messages of those types will accumulate in the inbox until it is
+        destroyed. You can check for them at any time.
+        """
+        connection = MessageBusConnection()
+        # We call this private method, really we mean this to be module scope.
+        connection._set_bus_and_message_types(self, wanted_types)
+        return connection
+
+    def outbox(self) -> 'MessageOutbox':
+        """
+        Get a connection object that only allows sending messages.
+        """
+        connection = MessageOutbox()
+        connection._set_bus(self)
+        return connection
+
+    def connect_output_file(self, file_path: str) -> Any:
+        """
+        Send copies of all messages to the given output file.
+
+        Returns opaque connection data which must be kept alive for the
```

Comment:
I'm not entirely sure what opaque means in this context

Reply:
I can explain more here.